### PR TITLE
feat(swapper): add dynamic slippage

### PR DESCRIPTION
**Pull request by 0xApotheosis**
- Add `recommendedSlippage` as an optional property of `TradeQuote`. This will be consumed by `web` when fetching the quote and used to set the local state slippage value, which is then passed into `buildTrade`. This architecture allows us to override the `recommendedSlippage` with advanced trade settings in the future, as well as display the recommended slippage in the UI.
- Implemented `recommendedSlippage` for Thorswap. The logic handles the liquidity impact to pools when trading RUNE assets (one hop) and none-RUNE assets (two hops).

Contributes to https://github.com/shapeshift/lib/issues/1094

Unblocks https://github.com/shapeshift/web/pull/3419

### CODE REVIEW

File: packages/swapper/src/api.ts [hunk 1]
```diff
@@ -130,8 +130,9 @@ export interface TradeQuote<C extends ChainId> extends TradeBase<C> {
   allowanceContract: string
   minimumCryptoHuman: string
   maximum: string
+  recommendedSlippage?: string
 
-  /** @deprecated Use sellAmountCryptoHuman instead */
+  /** @deprecated Use minimumCryptoHuman instead */
```
Comment (0xApotheosis) [inline packages/swapper/src/api.ts:135]: An unrelated typo I noticed whilst in the neighbourhood.

File: packages/swapper/src/api.ts [hunk 2]
```diff
@@ -284,5 +285,8 @@ export interface Swapper<T extends ChainId> {
    */
   filterAssetIdsBySellable(assetIds: AssetId[]): AssetId[]
 
+  /**
+   * Get transactions related to a trade
+   */
```
Comment (0xApotheosis) [inline packages/swapper/src/api.ts:290]: An unrelated fix whilst in the neighbourhood.

File: packages/swapper/src/swappers/thorchain/utils/getLimit/getLimit.test.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ describe('getLimit', () => {
       buyAssetTradeFeeUsd: '6.2049517907881932',
     }
     const limit = await getLimit(getLimitArgs)
-    expect(limit).toBe('580583')
+    expect(limit).toBe('577495')
```
Comment (0xApotheosis) [inline packages/swapper/src/swappers/thorchain/utils/getLimit/getLimit.test.ts:54]: These are slightly less than they used to be due to the addition of `ALLOWABLE_MARKET_MOVEMENT`.

File: packages/swapper/src/swappers/utils/constants.ts [hunk 4]
```diff
@@ -1,2 +1,3 @@
 export const APPROVAL_GAS_LIMIT = '100000' // Most approvals are around 40k, we've seen 72k in the wild, so 100000 for safety.
 export const DEFAULT_SLIPPAGE = '0.03' // 3%
+export const ALLOWABLE_MARKET_MOVEMENT = '0.005' // 0.5% is what https://app.thorswap.finance/ uses
```
Comment (0xApotheosis) [inline packages/swapper/src/swappers/utils/constants.ts:3]: Now that our THORChain slippage percentage values are so tight (accurate) we need to account for market movements.

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getSlippage.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+type GetSlippageArgs = {
+  inputAmountThorPrecision: BN
+  daemonUrl: string
+  buyAssetId: AssetId
+  sellAssetId: AssetId
+}
+
+export const getSlippage = async ({
```
Comment (0xApotheosis) [inline packages/swapper/src/swappers/thorchain/getThorTradeQuote/getSlippage.ts:60]: This guy is the meat of this PR.

**Review by 0xApotheosis — COMMENTED** [review]
Notes for reviewers.

**Review by gomesalexandre — COMMENTED** [review]
Looks good, first pass at logic and runtime, tested with linked lib in the matching web PR, confirmed `getSlippage` fires an XHR to Thor API and the slippage looks good

image
image

As can be seen on the screnshot above, on-chain memo includes the correct slippage, which is reflected on the web side when comparing against tests:

image

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts [hunk 6]
```diff
@@ -10,12 +10,11 @@ import { thorService } from '../thorService'
...
-const getSwapOutput = (inputAmount: BN, pool: ThornodePoolResponse, toRune: boolean): BN => {
-  const x = inputAmount
-  const X = toRune ? pool.balance_asset : pool.balance_rune
-  const Y = toRune ? pool.balance_rune : pool.balance_asset
-  const numerator = bn(x).times(X).times(Y)
-  const denominator = bn(x).plus(X).pow(2)
+export const getSwapOutput = (inputAmount: BN, pool: ThornodePoolResponse, toRune: boolean): BN => {
+  const inputBalance = toRune ? pool.balance_asset : pool.balance_rune
+  const outputBalance = toRune ? pool.balance_rune : pool.balance_asset
+  const numerator = bn(inputAmount).times(inputBalance).times(outputBalance)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts]: Right you are! Addressed in https://github.com/shapeshift/lib/pull/1102/commits/8e3bab8e4073657544c7f2d6a28dd7a2091009fe

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getSlippage.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  inputAmountThorPrecision,
+  daemonUrl,
+  buyAssetId,
+  sellAssetId,
+}: GetSlippageArgs): Promise<BN> => {
+  const { data: poolData } = await thorService.get<ThornodePoolResponse[]>(
+    `${daemonUrl}/lcd/thorchain/pools`,
+  )
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getSlippage.ts]: For sure - wrapped in https://github.com/shapeshift/lib/pull/1102/commits/c2d7cd2b7782cd4d4e562373f6ecea3ab4dc2aa9.

File: packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts [hunk 8]
```diff
@@ -115,6 +128,7 @@ export const getThorTradeQuote: GetThorTradeQuote = async ({ deps, input }) => {
       sellAsset,
       bip44Params,
       minimumCryptoHuman: minimumSellAssetAmountPaddedCryptoHuman,
+      recommendedSlippage: recommendedSlippage.toString(),
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/swapper/src/swappers/thorchain/getThorTradeQuote/getTradeQuote.ts]: Indeed! Addressed in https://github.com/shapeshift/lib/pull/1102/commits/fc3f6acb98b1a713ea780f5d2ffd354c2f922946.

**Review by gomesalexandre — APPROVED** [review]
All comments have been tackled, LGTM 🧠
